### PR TITLE
Meta data dictionary move and copy on write

### PR DESCRIPTION
**Pull request by blowekamp**

### CODE REVIEW

**Comment by blowekamp** [comment]
As noted by @N-Dekker the `operator[](const std::string & key) const` method was and is more problematically a violation of a constant interface because if key is not in the dictionary it may add a default value at the specified key.

I can not think of a good option to fix this interface.

File: Modules/Core/Common/src/itkMetaDataDictionary.cxx [hunk 1]
```diff
@@ -182,9 +191,32 @@ void
 MetaDataDictionary
 ::Clear()
 {
+  MakeUnique();
```
Comment (dzenanz) [inline Modules/Core/Common/src/itkMetaDataDictionary.cxx]: Will this create a copy, only to be erased by the next line?

Comment (blowekamp) [reply Modules/Core/Common/src/itkMetaDataDictionary.cxx]: I’ll look into that

**Comment by blowekamp** [comment]
This should be a WIP

File: Modules/Core/Common/src/itkMetaDataDictionary.cxx [hunk 2]
```diff
@@ -53,29 +52,36 @@ void
...
   return ( *m_Dictionary )[key];
 }
 
 const MetaDataObjectBase *
 MetaDataDictionary
 ::operator[](const std::string & key) const
 {
-  MetaDataObjectBase::Pointer entry = ( *m_Dictionary )[key];
-  const MetaDataObjectBase *  constentry = entry.GetPointer();
+  auto iter = m_Dictionary->find(key);
```
Comment (blowekamp) [inline Modules/Core/Common/src/itkMetaDataDictionary.cxx]: @N-Dekker I pondered deprecating this method, but instead I have just made it return a nullptr, and not modify the std::map when not does not exist. This const method does not exist in the std container, and the behavior is different than the non-constant method.

**Pull request merged**